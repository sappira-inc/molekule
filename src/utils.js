import { get, kebabCase } from 'lodash';
import styled from 'styled-components';

export const themeGet = (theme, lookup, fallback) => get(theme, lookup, fallback);

export const getComponentVariant = (theme, componentName, variant) => {
  const config = themeGet(
    theme,
    `components.${componentName}.variants.${variant}`,
    theme.variants[componentName][variant]
  );
  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentStyle = componentName => ({ theme }) => themeGet(theme, `components.${componentName}.style`);

export const getComponentClassName = ({ className, theme: { classPrefix }, variant }, name) =>
  `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

export const createComponent = ({ name, tag = 'div', as, style, props: defaultProps = () => ({}) }) => {
  const component = as ? styled(as) : styled[tag];

  return component.attrs(props => ({
    ...defaultProps(props),
    ...props,
    className: getComponentClassName(props, kebabCase(name)),
  }))`
    ${style}
    ${getComponentStyle(name)}
    ${({ styles = {} }) => styles[name] || {}}
  `;
};
