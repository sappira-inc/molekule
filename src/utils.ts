import kebabCase from 'lodash/kebabCase';
import get from 'lodash/get';
import styled, { CSSProp } from 'styled-components';

export const themeGet = (lookup: any, fallback?: any) => ({ theme }: any = {}) => get(theme, lookup, fallback);

export const getComponentVariant = (theme: any, componentName: string, variant: string) => {
  const config = themeGet(`variants.${componentName}.${variant}`)({
    theme,
  });

  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentSize = (theme: any, componentName: string, size: string) =>
  themeGet(`sizes.${componentName}.${size}`, {})({ theme });

export const getComponentStyle = (componentName: string) => themeGet(`styles.${componentName}`, {});

export const getVariantStyles = (componentName: string, variant: string) =>
  themeGet(`variants.${componentName}.${variant}.style`, {});

interface GetComponentClassNameProps {
  className: string;
  theme: {
    classPrefix: string;
  };
  variant: string;
}

const getComponentClassName = (
  { className, theme: { classPrefix }, variant }: GetComponentClassNameProps,
  name: string
) => `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

interface CreateComponentProps<T> {
  name: string;
  tag?: keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>;
  as?: React.ComponentType<any>;
  style?: ((props: any) => CSSProp) | CSSProp;
  props?: (props: T) => any;
}

export const createComponent = <
  T extends object,
  O extends keyof JSX.IntrinsicElements | React.ComponentType<any> = 'div'
>({
  name,
  tag = 'div',
  as,
  style,
  props: getBaseProps = () => ({}),
}: CreateComponentProps<T>) => {
  const component = styled<O>((as || tag) as any);

  return component.attrs<T>((props: any) => {
    const baseProps = getBaseProps(props);
    const finalProps = {
      ...baseProps,
      ...props,
      className: baseProps.className || props.className,
    };

    return {
      ...finalProps,
      className: getComponentClassName(finalProps, kebabCase(name)),
    };
  })<T>`
  ${style}
  ${getComponentStyle(name)}
  ${(p: any) => getVariantStyles(name, p.variant)}
  ${({ styles = {} }) => styles[name] || {}}
`;
};

// eslint-disable-next-line
const focusableFilter = (node: any) =>
  node.tabIndex >= 0 && !node.hasAttribute('disabled') ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_SKIP;

type FilterFn = (node?: any) => any;

const createTreeWalker = (root: any, currentNode: any, filterFn: FilterFn = () => NodeFilter.FILTER_ACCEPT) => {
  const treeWalker = document.createTreeWalker(
    root,
    NodeFilter.SHOW_ELEMENT,
    {
      acceptNode: filterFn,
    },
    false
  );
  treeWalker.currentNode = currentNode;
  return treeWalker;
};

export const findPreviousFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.previousNode() || treeWalker.currentNode === root) {
    return treeWalker.lastChild();
  }

  return treeWalker.currentNode;
};

export const findNextFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.nextNode() || treeWalker.currentNode === root) {
    treeWalker.currentNode = root;
    return treeWalker.firstChild();
  }

  return treeWalker.currentNode;
};

export const getNextCursorPosition = (cursorPos: any, oldValue: any, newValue: any) => {
  const cursorDiff = newValue.length - oldValue.length;
  let nextPosition = cursorPos;

  if (cursorDiff > 1 || cursorDiff < -1) {
    nextPosition += cursorDiff < 0 ? cursorDiff + 1 : cursorDiff - 1;
  }

  return Math.max(nextPosition, 0);
};
