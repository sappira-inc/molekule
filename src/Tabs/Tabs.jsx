import React, { Component } from 'react';
import { css } from 'styled-components';
import PropTypes from 'prop-types';
import { createComponent } from '../utils';

const TabsProvider = createComponent({
  name: 'Tabs',
  style: ({ vertical }) => css`
    display: flex;
    flex-direction: ${vertical ? 'row' : 'column'};
  `,
});

const TabList = createComponent({
  name: 'TabList',
  tag: 'ul',
  style: ({ vertical }) => css`
    display: flex;
    margin: 0;
    padding: 0;
    list-style: none;
    flex-direction: ${vertical ? 'column' : 'row'};

    ${!vertical &&
      css`
        border-bottom: 1px solid #dfe3e8;
      `};
  `,
});
const TabListItem = createComponent({
  name: 'TabListItem',
  tag: 'li',
  style: () => css`
    display: flex;
    margin: 0;
    padding: 0;
  `,
});

const Tab = createComponent({
  name: 'Tab',
  tag: 'button',
  style: ({ vertical, disabled }) => css`
    appearance: none;
    margin: 0;
    padding: 0;
    background: none;
    border: none;
    font-size: 14px;
    font-family: inherit;
    position: relative;
    justify-content: center;
    cursor: pointer;
    margin-${vertical ? 'left' : 'top'}: 1px;
    margin-${vertical ? 'right' : 'bottom'}: -1px;
    padding: ${vertical ? '0' : '0 12px'};
    outline: none;
    white-space: nowrap;
    opacity: ${disabled ? 0.5 : 1};

    &:active {
      outline: none;
    }
  `,
});

const TabTitle = createComponent({
  name: 'TabTitle',
  tag: 'span',
  style: ({ active, vertical, theme }) => css`
    display: block;
    color: ${active ? '#212b36' : '#637381'};
    padding: ${vertical ? '8px 12px 8px 8px' : '12px 8px'};
    border-${vertical ? 'left' : 'bottom'}: 4px solid ${active ? theme.colors.primary : 'transparent'};

    &::hover {
      color: #212b36;
      border-color: ${active ? theme.colors.primary : '#d9dee3'};
    }
  `,
});

const TabContent = createComponent({
  name: 'TabContent',
  tag: 'div',
  style: css`
    width: 100%;
  `,
});

class Tabs extends Component {
  static propTypes = {
    tabs: PropTypes.arrayOf(PropTypes.shape()).isRequired,
    active: PropTypes.number,
    vertical: PropTypes.bool,
    onChange: PropTypes.func,
  };

  static defaultProps = {
    vertical: false,
  };

  state = {
    active: this.props.active || 0,
  };

  componentDidUpdate() {
    if (this.props.active && this.props.active !== this.state.active) {
      this.setState({
        active: this.props.active,
      });
    }
  }

  handleTabClick = selected => {
    if (this.props.onChange) {
      this.props.onChange(selected);
    }

    this.setState({
      active: selected,
    });
  };

  renderHeadings() {
    return this.props.tabs.map((tab, i) => {
      const id = tab.id || i;

      return (
        <TabListItem key={id}>
          <Tab
            id={id}
            disabled={tab.disabled}
            vertical={this.props.vertical}
            onClick={() => {
              if (tab.onActive) {
                tab.onActive();
              }
              this.handleTabClick(i);
            }}>
            <TabTitle active={this.state.active === i} vertical={this.props.vertical}>
              {tab.title}
            </TabTitle>
          </Tab>
        </TabListItem>
      );
    });
  }

  renderContent() {
    const { content } = this.props.tabs[this.state.active];
    if (React.isValidElement(content)) {
      return content;
    }
    throw new Error('Tab content must be a valid React element.');
  }

  render() {
    return (
      <TabsProvider vertical={this.props.vertical}>
        <TabList vertical={this.props.vertical}>{this.renderHeadings()}</TabList>
        <TabContent>{this.renderContent()}</TabContent>
      </TabsProvider>
    );
  }
}

export default Tabs;
