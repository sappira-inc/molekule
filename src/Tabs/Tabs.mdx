---
menu: Core
name: Tabs
---

import { Playground, PropsTable } from 'docz'
import Tabs from './Tabs'

# Tabs

Tabs enable a user to quickly switch between several panels. They can be laid out either horizontally or vertically depending on the `vertical` prop.

## Layout

### Horizontal

<Playground>
  <Tabs
    tabs={[
      {
        id: 'test-a-one',
        title: 'Tab One',
        content: <div>Hey, I'm the first panel</div>,
      },
      {
        id: 'test-a-two',
        title: 'Tab Two',
        content: <div>Hey, I'm the second panel</div>,
      }, {
        id: 'test-a-three',
        title: 'Tab Three',
        content: <div>Hey, I'm the third panel</div>,
        disabled: true
      }]}
  />
</Playground>

### Vertical

<Playground>
  <Tabs
    vertical
    tabs={[
      {
        id: 'test-b-one',
        title: 'Tab One',
        content: <div>Hey, I'm the first panel</div>,
      },
      {
        id: 'test-b-two',
        title: 'Tab Two',
        content: <div>Hey, I'm the second panel</div>,
      }, {
        id: 'test-b-three',
        title: 'Tab Three',
        content: <div>Hey, I'm the third panel</div>,
        disabled: true
      }]}
  />
</Playground>

## Callbacks

### `onChange`
Tabs, by default, are self-contained and will control themselves. However, if an `onChange` prop is provided, Tabs can now be controlled by the `active`.

### `onActive`
`onActive` will be fired when a tab becomes active. Usually this is used to activate some async action to load data or for analytics.
