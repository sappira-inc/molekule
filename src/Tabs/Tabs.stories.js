import React from 'react';
import Tabs from './Tabs';

export default {
  title: 'Components|Tabs',
  component: Tabs,
};

export const Horizontal = () => (
  <>
    <Tabs
      tabs={[
        {
          id: 'test-a-one',
          title: 'Tab One',
          content: <div>Hey, first panel</div>,
        },
        {
          id: 'test-a-two',
          title: 'Tab Two',
          content: <div>Hey, second panel</div>,
        },
        {
          id: 'test-a-three',
          title: 'Tab Three',
          content: <div>Hey, third panel</div>,
          disabled: true,
        },
      ]}
    />
  </>
);

export const Vertical = () => (
  <Tabs
    vertical
    tabs={[
      {
        id: 'test-b-one',
        title: 'Tab One',
        content: <div>Hey, first panel</div>,
      },
      {
        id: 'test-b-two',
        title: 'Tab Two',
        content: <div>Hey, second panel</div>,
      },
      {
        id: 'test-b-three',
        title: 'Tab Three',
        content: <div>Hey, third panel</div>,
        disabled: true,
      },
    ]}
  />
);
