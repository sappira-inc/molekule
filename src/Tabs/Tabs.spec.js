import React from 'react';
import { renderWithTheme } from '../../test/utils';
import Tabs from './Tabs';

const TabContent = ({ body }) => <div>This is the content for Tab {body}</div>;
const onActiveSpy = jest.fn();
const TABS = [
  {
    id: 'test-one',
    title: 'Tab One',
    content: <TabContent body="one" />,
  },
  {
    id: 'test-two',
    title: 'Tab Two',
    content: <TabContent body="two" />,
    onActive: onActiveSpy,
  },
];

describe('<Tabs />', () => {
  afterEach(() => {
    onActiveSpy.mockClear();
  });

  test('renders', () => {
    const component = renderWithTheme(<Tabs tabs={TABS} />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
