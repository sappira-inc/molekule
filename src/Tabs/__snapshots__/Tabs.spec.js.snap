// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> renders 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  padding: 0;
  list-style: none;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  border-bottom: 1px solid #F4F4F4;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  padding: 0;
}

.c3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0;
  padding: 0;
  background: none;
  border: none;
  font-size: 14px;
  font-family: inherit;
  position: relative;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  margin-top: 1px;
  margin-bottom: -1px;
  padding: 0 12px;
  outline: none;
  white-space: nowrap;
  opacity: 1;
}

.c3:active {
  outline: none;
}

.c4 {
  display: block;
  color: #494D55;
  padding: 12px 8px;
  border-bottom: 4px solid #0958F3;
}

.c4::hover {
  color: #212b36;
  border-color: #0958F3;
}

.c5 {
  display: block;
  color: #A4A6AA;
  padding: 12px 8px;
  border-bottom: 4px solid transparent;
}

.c5::hover {
  color: #212b36;
  border-color: #E8E9EA;
}

.c6 {
  width: 100%;
}

<div
    class="c0 re-tabs"
  >
    <ul
      class="c1 re-tab-list"
    >
      <li
        class="c2 re-tab-list-item"
      >
        <button
          class="c3 re-tab"
          id="test-one"
        >
          <span
            class="c4 re-tab-title"
          >
            Tab One
          </span>
        </button>
      </li>
      <li
        class="c2 re-tab-list-item"
      >
        <button
          class="c3 re-tab"
          id="test-two"
        >
          <span
            class="c5 re-tab-title"
          >
            Tab Two
          </span>
        </button>
      </li>
    </ul>
    <div
      class="c6 re-tab-content"
    >
      <div>
        This is the content for Tab one
      </div>
    </div>
  </div>
</DocumentFragment>
`;
